src/app.js

```diff
@@ -61,24 +62,65 @@ async function displayTrace(assetUrl, fileData) {
 
   /**
-  * For the object data URL (in getAssetUrl) we just encode the traces/traceid path once. but here we do it TWICE. Why????!
-  * Because DevTools incorrectly double decodes the path.
-  *  - First decode in `Runtime.Runtime.queryParam()` with the searchParams.get call
-  *  - Second decode in TimelinePanel's `handleQueryParam()` for loadTimelineFromURL it happens again.
-  * TODO: fix that bug in devtools.
-  */
+   * For the object data URL (in getAssetUrl) we just encode the traces/traceid path once. but here we do it TWICE. Why????!
+   * Because DevTools incorrectly double decodes the path.
+   *  - First decode in `Runtime.Runtime.queryParam()` with the searchParams.get call
+   *  - Second decode in TimelinePanel's `handleQueryParam()` for loadTimelineFromURL it happens again.
+   * TODO: fix that bug in devtools.
+   */
   const encodedAssetUrl = encodeURIComponent(assetUrl);
   const hostedDtViewingTraceUrl = new URL(devtoolsBaseUrl);
   hostedDtViewingTraceUrl.searchParams.set('loadTimelineFromURL', encodedAssetUrl);
 
   console.log('Trace opening in DevTools…', filename);
-  const iframe = $('iframe#ifr');
+  const iframe = $('iframe#ifr-dt');
   iframe.onload = _ => {
-    // Technically devtools iframe just loaded (didnt 404). We assume the trace loaded succfessully too. 
+    // Technically devtools iframe just loaded (didnt 404). We assume the trace loaded succfessully too.
     // Can't really extract errors from that iframe.....
     console.log('Trace loaded.', filename, 'Uploaded:', dateStr);
-  }
+  };
   iframe.src = hostedDtViewingTraceUrl.href;
+
+  // Warm up perfetto iframe
+  const iframePerfetto = $('iframe#ifr-perfetto');
+  iframePerfetto.src = 'https://ui.perfetto.dev/';
+  // global vars for the perfetto load... it's gross. i'm sorry.
+  globalThis.traceAssetUrl = assetUrl;
+  globalThis.traceTitle = `${filename} — ${dateStr}`;
+}
+
+// https://perfetto.dev/docs/visualization/deep-linking-to-perfetto-ui
+async function showTraceInPerfetto(iframePerfetto, traceAssetUrl, traceTitle) {
+  const ORIGIN = 'https://ui.perfetto.dev';
+  const timer = setInterval(() => iframePerfetto.contentWindow.postMessage('PING', ORIGIN), 50);
+
+  const onPerfettoMsg = async evt => {
+    if (evt.data !== 'PONG') return;
+    // We got a PONG, the UI is ready.
+    window.clearInterval(timer); window.removeEventListener('message', onPerfettoMsg);
+
+    const traceBuffer = await fetch(traceAssetUrl).then(r => r.arrayBuffer());
+    // Send trace
+    const payload = {
+      perfetto: {
+        buffer: traceBuffer,
+        title: traceTitle,
+      },
+    };
+    iframePerfetto.contentWindow.postMessage(payload, ORIGIN);
+    iframePerfetto.classList.add('traceloaded');
+  };
+
+  window.addEventListener('message', onPerfettoMsg);
+}
+
+function togglePerfettoViz(){
+  const iframePerfetto = $('iframe#ifr-perfetto');
+  const shouldShowPerfetto = iframePerfetto.classList.toggle('visible');
```

Comment:
Is this meant to write/toggle state here or `.contains` is the right use?

Reply:
good catch i'm relying on the return value of `toggle()` which.. i had to look up.. it's defined.  :)

![image](https://github.com/paulirish/trace.cafe/assets/39191/ab54ab1e-d161-4b0b-a9e2-b2e188f565bf)

kinda weird logically to read it. 
but.. beats a toggle followed by a contains?